fs: support io_uring in `fs::write`

mox692 opened the pull request:
Add io-uring version of `fs::write`. I think the change itself is mostly straightforward. This PR might need to coordinate with https://github.com/tokio-rs/tokio/pull/7547 regarding the usage of cfg.

Related issue: https://github.com/tokio-rs/tokio/issues/7266

Darksonn commented on tokio/src/fs/write.rs:
```diff
@@ -24,6 +24,49 @@ use std::{io, path::Path};
+}
+
+#[cfg(all(tokio_uring, feature = "rt", feature = "fs", target_os = "linux"))]
+async fn write_uring(path: impl AsRef<Path>, mut buf: &[u8]) -> io::Result<()> {
```
I would change both methods to take `&Path` and `crate::util::as_ref::OwnedBuf`, to avoid generics in the actual implementations.

Darksonn commented on tokio/src/io/uring/write.rs:
```diff
@@ -0,0 +1,42 @@
+            .offset(offset)
+            .build();
+
+        // SAFETY: `fd` and `buf` is valid until this future completes.
+        // If this operation is cancelled, drop of `Op` occurs before dropping
+        // `fd` or `buf`, ensuring the safety contract holds.
+        let op = unsafe { Op::new(sqe, Write) };
```
This logic I don't buy. Using `mem::forget` on the future it's possible to cause UAF. Some form of ownership is required.

mox692 commented:
I made some changes in [da5b9e1](https://github.com/tokio-rs/tokio/pull/7567/commits/da5b9e192d376277c2892718771fd6b9d7a5aa3d):

1. Fixed to rely on an owned buffer.
2. Made the `Write` struct own resources:
    - Introduced & updated some ref-counted owned utility types such as `SharedFd` and `OwnedBuf`.

I think (2) is not strictly required (it's more of a coding style change), but makes it clear that an operation owns the resources when reading the code IMO.

mox692 commented on tokio/src/fs/write.rs:
```diff
@@ -24,8 +25,49 @@ use std::{io, path::Path};
+        if n == 0 {
+            return Err(io::ErrorKind::WriteZero.into());
+        }
+        buf.advance(n);
```
> advance is pretty expensive

Is it? I'm assuming it's just a cheap numeric operation, unless I'm missing something.

https://github.com/tokio-rs/tokio/blob/8bb7cac95d900d878e4d76484c6b6f4acda810fb/tokio/src/util/as_ref.rs#L55

Darksonn replied:
Oh I see, I missed you did it like that. In that case, the expensive operation is converting from `Vec<u8>` to `Arc<[u8]>`. I don't think we want those refcounted types if we can at all avoid it.

mox692 commented on tokio/src/io/uring/write.rs:
```diff
@@ -0,0 +1,40 @@
+            .offset(offset)
+            .build();
+
+        // SAFETY: `fd` and `buf` are owned by caller function, ensuring these params are
+        // valid until operation completes.
+        let op = unsafe { Op::new(sqe, Write) };
```
If the future returned by `fs::write` is dropped (i.e. cancelled), `Op` gets dropped first and then `OwnedBuf`, since `OwnedBuf` is declared first in the `write_uring`. This follows the drop order [rules](https://doc.rust-lang.org/reference/destructors.html#r-destructors.scope.intro) of the local variables. (I also checked the order locally)

```rust
pub async fn write() -> io::Result<()> {
  let owned_buf = upgrade();
  ...
  write_uring(owned_buf).await
}

async fn write_uring(owned_buf: OwnedBuf) -> io::Result<()> {
  ...
  let op = Op::write_at(); // Op is declared after OwnedBuf
  op.await
}
```

It's probably worth adding a short comment about this, in addition to making the `write_at` unsafe.

mox692 replied:
Made `write_at` unsafe and add more safety comments.

https://github.com/tokio-rs/tokio/pull/7567/commits/05c8c17e06ca243dadd44d1467447fe6e4ed8dc9

Darksonn replied:
I'm pretty sure that when you cancel the write operation, the buffer needs to be stored here:

https://github.com/tokio-rs/tokio/blob/86400a1920c4caef28ae5178af0fcf1a34822b92/tokio/src/runtime/driver/op.rs#L31-L35

so that the buffer stays alive until this line:

https://github.com/tokio-rs/tokio/blob/86400a1920c4caef28ae5178af0fcf1a34822b92/tokio/src/runtime/io/driver/uring.rs#L93-L97

Just because you submitted a cancel operation to the driver, you can't be sure that it has been cancelled *yet* even if it will be soon, and I don't think it's possible to synchronously wait for cancellation in a destructor.

Darksonn commented on tokio/src/fs/write.rs:
```diff
@@ -24,8 +24,59 @@ use std::{io, path::Path};
+    while offset < total {
+        // There is a cap on how many bytes we can write in a single uring write operation.
+        // ref: https://github.com/axboe/liburing/discussions/497
+        let len = std::cmp::min(total - offset, u32::MAX as usize) as u32;
```
Could also make sense to move this logic inside `Op::write_at` and have it accept usize. After all, it's indistinguishable to the caller from any other situation that writes the bytes partially. That way it's more natural for the caller as the caller expects usize for both input and output length.
```suggestion
        let len = u32::try_from(total - offset).unwrap_or(u32::MAX);
```

Darksonn commented on tokio/src/fs/write.rs:
```diff
@@ -24,8 +24,59 @@ use std::{io, path::Path};
+        .expect("unexpected in-flight operation detected")
+        .into();
+
+    let total: usize = buf.len();
+    let mut offset: usize = 0;
+    while offset < total {
```
Since `write_at` accepts u64 for file positions, let's use that here too.

Darksonn commented on tokio/src/io/uring/write.rs:
```diff
@@ -0,0 +1,58 @@
+        len: u32,
+        file_offset: u64,
+    ) -> io::Result<Self> {
+        // Check if `buf_offset` stays in bounds of the allocation
+        debug_assert!(buf_offset + len as usize <= buf.len());
+
+        // SAFETY:
+        // - `buf_offset` stays in bounds of the allocation.
+        // - `buf` is derived from an actual allocation, and the entire memory
+        //    range is in bounds of that allocation.
+        let ptr = unsafe { buf.as_ptr().add(buf_offset) };
```
```suggestion
        let ptr = buf[buf_offset .. buf_offset + len].as_ptr();
```

Darksonn commented on tokio/src/runtime/driver/op.rs:
```diff
@@ -16,6 +17,11 @@ pub(crate) enum CancelData {
+    Write(
+        // This field isn't accessed directly, but it holds cancellation data,
+        // so `#[allow(dead_code)]` is needed.
+        #[allow(dead_code)] Write,
```
Let's move this allow to the entire enum to avoid repetition.

Darksonn commented on tokio/src/util/as_ref.rs:
```diff
@@ -7,6 +7,22 @@ pub(crate) enum OwnedBuf {
     Bytes(bytes::Bytes),
 }
 
+impl OwnedBuf {
+    cfg_tokio_uring! {
+        pub(crate) fn len(&self) -> usize {
+            match self {
+                Self::Vec(vec) => vec.len(),
+                #[cfg(feature = "io-util")]
+                Self::Bytes(bytes) => bytes.len(),
+            }
+        }
+
+        pub(crate) fn as_ptr(&self) -> *const u8 {
```
I think all of this can go through the `as_ref()` impl.

Darksonn reviewed (approved):
Looks good to me!
